Let `DB.readTxHistory` take `SortOrder` and `Range SlotId`

Anviking opened the pull request:
# Issue Number

#466 


# Overview

- [x] @Anviking I added `SortDirection` and `Filter a` (for imposing `>=`/`<=` requirements) to core
- [x] @Anviking I made `DB.readTxHistory` take and apply `sortDir` and `filter` with implementations in `Sqlite`, `MVar` and Sqlite's StateMachine tests.
- [x] @Anviking Some final polish / decisions missing. Done I think.
- [x] @jonathanknowles: Replace usage of `Test.QuickCheck.Instances.Time` with `UniformTime` from module `Test.Utils.Time`. 
- [x] @jonathanknowles: Use `arbitraryBoundedEnum` to avoid repeating the definition of `SortOrder`.

# Comments

<!-- Additional comments or screenshots to attach if any -->



<!-- 
Don't forget to:

 ✓ Self-review your changes to make sure nothing unexpected slipped through
 ✓ Assign yourself to the PR
 ✓ Assign one or several reviewer(s)
 ✓ Once created, link this PR to its corresponding ticket
 ✓ Acknowledge any changes required to the Wiki
-->

Anviking committed:
Support `Range` and `SortOrder` in DB.readTxHistory

I made DB.readTxHistory take and apply sortOrder and range filtering with implementations in Sqlite, MVar and Sqlite's StateMachine tests.

Squashed from:

Ensure readTxHistory tests pass with `Descending noFilter`
[WIP] Make tests concider sorting and filtering
Do actual Sqlite sorting and filtering
Apply suggestions from code review
Introduce defaultTxSortDirection
Temporary overflow fix to make StateMachine tests pass
Rename Filter to Range`
Fixup: more polish

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -507,6 +511,23 @@ data TransactionInfo = TransactionInfo
+    deriving (Eq, Show)
+
+-- Represents the range [start, end]
+data Filter a = Filter
```
Should add some clearer comment here

KtorZ replied:
Ah! This si quite similar to what I proposed here: https://github.com/input-output-hk/cardano-wallet/pull/586#discussion_r308272438. I like it :)

KtorZ commented on lib/core/src/Cardano/Wallet/DB/MVar.hs:
```diff
@@ -125,10 +130,16 @@ newDBLayer = do
+        , readTxHistory = \key sortDir filt -> let
+                sort' = case sortDir of
+                    Ascending -> id
+                    Descending -> reverse
```
What about:

```hs
direction = case sortDir of
    Ascending -> id
    Descending -> Down

{- ... -}

result = filter ... 
    . sortOn slot . Map.toList . txHistory
slot = direction . slotId . snd . snd
```

to avoid traversing the whole list thrice :s 
Though I just noticed it's the MVar implementation ... So, I guess we don't really care :sweat_smile: ...

Anviking replied:
I tried 
```haskell
direction = case sortDir of
    Ascending -> id
    Descending -> Down
```
but it doesn't type-check since `a -> a` and `a -> Down a` are different types.

> So, I guess we don't really care

That's what I thought too, however:

I think `sortBy` and `comparing` should work. I'll try that.

KtorZ replied:
Ah, yeay. Maybe `Down . Down` :man_shrugging: . But, doesn't matter much here. We should definitely remove that code...

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -347,9 +351,12 @@ newDBLayer logConfig trace fp = do
-              selectTxHistory @t wid []
+              selectTxHistory @t wid sortDir $ catMaybes
+                [ (TxMetaSlotId >=.) <$> W.filterStart filt
+                , (TxMetaSlotId <=.) <$> W.filterStart filt
```
`filterStart` -> `filterEnd` ?

Anviking replied:
Hmm… good catch. I just ran the *full* suite, locally and I'm glad it failed.

KtorZ commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -739,18 +746,22 @@ selectTxs txids = do
+  where
+    sortOpt = case dir of
+        W.Ascending -> [Asc TxMetaSlotId, Desc TxMetaTxId]
+        W.Descending -> [Desc TxMetaSlotId, Asc TxMetaTxId]
```
What's the rationale for also changing the ordering on `TxMetaTxId` ? This second ordering is there to keep the sorting stable when multiple transactions have the same slot id. I believe the SQL index is optimised for one particular direction, so I'd rather stick with only a single direction :+1:

Anviking replied:
I think that would break* `readTxHistory wid Descending noFilter === reverse (readTxHistory wid Ascending noFilter)` which I thought would be a nice property to have.

🤔 )* although I'm not certain that's the case now either

rvl replied:
Better move the comment about indexes down to the `where` clauses.
With sqlite, for sort speed, it doesn't matter which direction the index is created in, there just needs to be a sorted index.

Anviking replied:
Moved.

KtorZ commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -507,6 +511,23 @@ data TransactionInfo = TransactionInfo
     } deriving (Show, Eq, Ord)
 
+data SortDirection = Ascending | Descending
+    deriving (Eq, Show)
```
We are "hard-coding" `Descending` in quite a few places for being the default direction. What about having an explicit top-level identifier `defaultDirection = Descending` and use that `defaultDirection` everywhere else the semantic applies?

Anviking replied:
Great idea I think!

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -526,6 +534,18 @@ generator (Model _ wids) = Just $ frequency $ fmap (fmap At) <$> concat
     genPrivKey = elements ["pk1", "pk2", "pk3"]
 
+    genSortDir :: Gen SortDirection
+    genSortDir = QC.oneof [return Ascending, return Descending]
```
```suggestion
genSortDir = QC.elements [Ascending, Descending]
```

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -526,6 +534,18 @@ generator (Model _ wids) = Just $ frequency $ fmap (fmap At) <$> concat
+    genSortDir = QC.oneof [return Ascending, return Descending]
+
+    genFilter :: Gen (Filter SlotId)
+    genFilter = Filter <$> genSId <*> genSId
```
So this potentially generates invalid ranges :thinking: ?

Anviking replied:
Yes. This PR allows unsatisfiable filters like`x > 10 && x < 2`. 

I originally thought that would be fine… but maybe we should do something… 
1. `unsafeMkFilter :: Maybe a -> Maybe a -> Filter a -- throws` 
2. `mkFilter :: Maybe a -> Maybe a -> Maybe (Filter a)` 
3. `mkFilter :: Maybe a -> Maybe a -> Filter a -- may swap the arguments such that low <= high is always true`

But I don't really like any of them. Maybe we are fine with doing nothing.

Also I'm not sure I'm totally thrilled about simply calling this type `Filter`.

jonathanknowles replied:
> Also I'm not sure I'm totally thrilled about simply calling this type `Filter`.

Maybe just call it `Range`, given the comment:

```hs
-- Represents the range [start, end]
```

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DBSpec.hs:
```diff
@@ -66,8 +66,11 @@ import Cardano.Wallet.Primitive.Types
     , Hash (..)
     , SlotId (..)
+    , SortDirection
+    , SortDirection (..)
```
Double import here :point_up:

KtorZ commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -496,6 +500,18 @@ instance {-# OVERLAPS #-} Arbitrary (Key 'RootK XPrv, Passphrase "encryption")
+
+instance Arbitrary SortDirection where
+    shrink _ = []
+    arbitrary = oneof [ return Ascending, return Descending ]
```
```suggestion
arbitrary = elements [Ascending, Descending]
```

```suggestion
arbitrary = genericArbitrary
```

Anviking commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -526,6 +537,22 @@ generator (Model _ wids) = Just $ frequency $ fmap (fmap At) <$> concat
+        -- FIXME: Temporary to check if this is tha cause of StateMachine
+        -- failure
+        genInt :: Num a => Gen a
+        genInt = fromIntegral <$> arbitrary @Word32
```
Seems we need mind the ~48 bit limit on epochs from this hack https://github.com/input-output-hk/cardano-wallet/blob/29c9a6a9111d80d4874a63c07c2f138dfadae104/lib/core/src/Cardano/Wallet/DB/Sqlite/Types.hs#L231-L236

Anviking commented on lib/core/src/Cardano/Wallet.hs:
```diff
@@ -758,7 +761,9 @@ newWalletLayer tracer bp db nw tl = do
-        liftIO $ assemble tip <$> DB.readTxHistory db (PrimaryKey wid)
+        liftIO $ assemble tip
+            <$> DB.readTxHistory db (PrimaryKey wid)
+                defaultTxSortOrder noFilter
```
TODO: `_listTransactions` should take sort and filter parameters as well. 

`DB.readTxHistory` is well-tested through the `StateMachine` tests. Not sure we need to do anything in particular on the WalletLayer level 🤔

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -306,7 +308,8 @@ newDBLayer logConfig trace fp = do
                       utxo <- selectUTxO cp
-                      txs <- selectTxHistory @t wid [TxMetaStatus ==. W.Pending]
+                      txs <- selectTxHistory @t wid
+                        W.defaultTxSortOrder [TxMetaStatus ==. W.Pending]
```
Indentation here should be 4 spaces.

jonathanknowles commented on lib/core/src/Cardano/Wallet/DB/Sqlite.hs:
```diff
@@ -739,18 +745,22 @@ selectTxs txids = do
     -> [Filter TxMeta]
     -> SqlPersistT IO [(W.Hash "Tx", (W.Tx t, W.TxMeta))]
-selectTxHistory wid conditions = do
+selectTxHistory wid dir conditions = do
```
Just a small thing, but for a value of type `SortOrder`, perhaps it would be clear to call it `order`, or `sortOrder`?

This way, names of types and values are consistent.

(I can understand why this might have happened, as we introduced `SortOrder` and `SortDirection` at the same time.)

Anviking replied:
💯 thanks for bringing up!

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -326,6 +330,23 @@ instance ToText SortOrder where
     fromText = fromTextToBoundedEnum SnakeLowerCase
 
+defaultTxSortOrder :: SortOrder
+defaultTxSortOrder =  Descending
```
Just wondering, why is it better to have `Descending` as a default?

Anviking replied:
Hm indeed… this is currently not used from the API as a "defaultTxSortOrder".

It is currently used for:
1. Ordering the DB read for pending transactions when reading a wallet checkpoint. This appears **pointless** as it will be put in a `Set (Tx, TxMeta)` later. https://github.com/input-output-hk/cardano-wallet/blob/315fec4d4cb56114e1fd59eedd420ea963e3cf29/lib/core/src/Cardano/Wallet/DB/Sqlite.hs#L312
2. `DBSpec` which just needs an arbitrary sort-order.
3. `SqliteFileModeSpec` which just needs an arbitrary sort-order https://github.com/input-output-hk/cardano-wallet/blob/315fec4d4cb56114e1fd59eedd420ea963e3cf29/lib/core/test/unit/Cardano/Wallet/DB/SqliteFileModeSpec.hs#L137
4. `listTransactions` in `WalletLayer`, which is temporary

Maybe we should remove this from `Primitive.Types`.

Anviking replied:
Removed in #592 .

jonathanknowles replied:
If this `Descending` value is used as a default _somewhere_ (be it in the API or elsewhere), then I'd much rather see it defined as a named constant than as a magic value embedded in the functions that use it.

See: https://github.com/input-output-hk/cardano-wallet/pull/592#pullrequestreview-268762445

KtorZ replied:
> listTransactions in WalletLayer, which is temporary

What do you mean "temporary" :sweat_smile: ?

---

I think the rationale is: the ordering should come from the caller of the wallet layer. So, starting from the `WalletLayer`, the sort order shouldn't be `Maybe` but a plain value (because a default at this stage would be totally arbitrary). 
At the API-level however, we decide of a default ordering for which `Descending` makes sense (most recent results first).

Anviking replied:
> What do you mean "temporary" 😅 ?

😄 `listTransactions` used a hard-coded `Descending` order instead of its argument.

With a default value in `listTransactions` we do need to refer to `Descending`, which I didn't think about.

>  the sort order shouldn't be Maybe but a plain value (because a default at this stage would be totally arbitrary).

Yes. Unless one argues that defaulting to `Descending` shouldn't be unique to the API. Since you seem positive, I think I'll go with it though.

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -326,6 +330,23 @@ instance ToText SortOrder where
+data Filter a = Filter
+    { filterStart :: Maybe a
+    , filterEnd :: Maybe a
+    } deriving (Eq, Show)
```
Perhaps call this `Range`? After all, your comment states "Represents the **range**". :)

Other suggestions:

* `noFilter` -> `infiniteRange` or `openRange`.
* `matchesFilter` -> `isWithinRange` or `withinRange`.

Anviking replied:
👍 I went with `Range`, `isWithinRange` and `wholeRange`.

jonathanknowles replied:
I hadn't thought of `wholeRange`. Nice name!

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -326,6 +330,23 @@ instance ToText SortOrder where
+    } deriving (Eq, Show)
+
+noFilter :: Filter a
+noFilter = Filter Nothing Nothing
```
Perhaps:
```hs
infiniteRange :: Range a
infiniteRange = Range Nothing Nothing
```
Or:
```hs
openRange :: Range a
openRange = Range Nothing Nothing
```

Anviking replied:
https://github.com/input-output-hk/cardano-wallet/pull/588#discussion_r308716875

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -326,6 +330,23 @@ instance ToText SortOrder where
+matchesFilter :: Ord a => Filter a -> a -> Bool
+matchesFilter (Filter low high) x =
+    (maybe True (x >=) low) &&
+    (maybe True (x <=) high)
```
Perhaps:

```hs

withinRange :: Ord a => a -> Range a -> Bool
withinRange x (Range low high) =
    (maybe True (x >=) low) &&
    (maybe True (x <=) high)
```

Anviking replied:
👍  https://github.com/input-output-hk/cardano-wallet/pull/588/files#r308716875

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -250,12 +260,13 @@ mPutTxHistory wid txList m@(M cp metas txs pk)
-mReadTxHistory :: MWid -> MockOp TxHistory
-mReadTxHistory wid m@(M cp _ txs _)
+mReadTxHistory :: MWid -> SortOrder -> Filter SlotId -> MockOp TxHistory
+mReadTxHistory wid sortDir filt m@(M cp _ txs _)
```
Suggest `order` instead of `sortDir` here (for consistency with the type name).

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -526,6 +537,22 @@ generator (Model _ wids) = Just $ frequency $ fmap (fmap At) <$> concat
     genPrivKey = elements ["pk1", "pk2", "pk3"]
 
+    genSortDir :: Gen SortOrder
+    genSortDir = QC.elements [Ascending, Descending]
```
Suggestion: `genSortDir` -> `genSortOrder`.

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -526,6 +537,22 @@ generator (Model _ wids) = Just $ frequency $ fmap (fmap At) <$> concat
+            [ return Nothing
+            , Just <$> (SlotId <$> genInt <*> genInt)
+            ]
+        -- FIXME: Temporary to check if this is tha cause of StateMachine
```
typo: "tha" -> "the".

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DBSpec.hs:
```diff
@@ -186,10 +194,14 @@ once_ xs = void . once xs
+filterTxHistory :: SortOrder -> Filter SlotId -> TxHistory -> TxHistory
+filterTxHistory dir filt =
+    filter (matchesFilter filt . slotId . snd . snd)
+    . (if dir == Ascending then reverse else id)
```
Suggest `dir` -> `order`.

jonathanknowles commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -497,6 +502,18 @@ instance {-# OVERLAPS #-} Arbitrary (Key 'RootK XPrv, Passphrase "encryption")
+
+instance Arbitrary SortOrder where
+    shrink _ = []
+    arbitrary = genericArbitrary
```
Instead of `genericArbitrary` here, we can use `arbitraryBoundedEnum`.

jonathanknowles commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -497,6 +502,18 @@ instance {-# OVERLAPS #-} Arbitrary (Key 'RootK XPrv, Passphrase "encryption")
+
+instance Arbitrary UTCTime where
+    shrink _ = []
+    arbitrary = posixSecondsToUTCTime . toEnum <$> arbitrary
```
In other places, we generate arbitrary `UTCTime` values with:
```
lib/core/test/unit/Data/Time/TextSpec.hs
29:import Test.QuickCheck.Instances.Time

lib/core/test/unit/Cardano/Wallet/DBSpec.hs
143:import Test.QuickCheck.Instances.Time

lib/core/test/unit/Cardano/Wallet/Api/TypesSpec.hs
177:import Test.QuickCheck.Instances.Time
```
But perhaps your instance is nicer. Perhaps we should use the same instance in all places?

Anviking replied:
🤷‍♂ The `Test.QuickCheck.Instances.Time` seem to catch removing one of the bound-checks in `Sqlite.hs` — i.e. work fine — so I'll go with that for consistency.

jonathanknowles replied:
Great!

Having said that, I seem to remember that the `Test.QuickCheck.Instances.Time` instance generates dates that are biased toward a very long time ago. I think it would be worth investigating the distribution, and seeing if we need to improve on it.

Though that work is definitely outside the scope of this PR. I've added an item to the blackboard here: https://github.com/input-output-hk/cardano-wallet/wiki/Blackboard#investigate-the-arbitrary-instance-for-utctime-in-testquickcheckinstancestime

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -396,8 +407,8 @@ runIO db = fmap Resp . go
             catchNoSuchWallet Unit $ putTxHistory db (PrimaryKey wid) (Map.fromList txs)
-        ReadTxHistory wid ->
-            Right . TxHistory <$> readTxHistory db (PrimaryKey wid)
+        ReadTxHistory wid sortDir filt ->
```
Suggest `order` instead of `sortDir` here (for consistency with type name).

rvl reviewed (approved):
Looking good

Anviking committed:
Remove use of defaultTxSortOrder in SqliteFileModeSpec

Anviking committed:
Remove use of defaultTxSortOrder in DBSpec

Anviking committed:
Remove use of defaultTxSortOrder in SqliteSpec

Anviking committed:
Remove use of defaultTxSortOrder in Sqlite.hs and Primitive.Types

Anviking committed:
Move notion of defaultTxSortOrder from WalletLayer to Api

jonathanknowles commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -326,6 +330,39 @@ instance ToText SortOrder where
+-- - [start, end]  Range (Just start) (Just end)   \x -> x >= start && x <= end
+-- - [start,∞)     Range (Just start) Nothing      \x -> x >= start
+-- - (-∞,end)      Range Nothing (Just end)        \x -> x <= end
+-- - (-∞,∞)        Range Nothing Nothing           \_x -> True
```
Nice comment!

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DBSpec.hs:
```diff
@@ -186,10 +194,14 @@ once_ xs = void . once xs
+filterTxHistory :: SortOrder -> Range SlotId -> TxHistory -> TxHistory
+filterTxHistory order range =
+    filter (isWithinRange range . slotId . snd . snd)
+    . (if order == Ascending then reverse else id)
```
I reckon we ought to use a case statement here, similar to:

```hs
(case order of
    Ascending -> reverse
    Descending -> id)
```
With a case statement that checks all cases, if someone adds a new constructor later on (however unlikely), then the compiler will remind us to consider the new case at this location.

This is philosophically similar to: https://github.com/input-output-hk/cardano-wallet/wiki/Coding-Standards#avoid-wildcard-when-pattern-matching-on-adts

Anviking replied:
👍 Was afraid of the need for `()` and multiline, but it works totally fine.

KtorZ commented on lib/core/test/unit/Cardano/Wallet/DB/SqliteFileModeSpec.hs:
```diff
@@ -131,7 +133,9 @@ spec =  do
-            testOpeningCleaning f (`readTxHistory` testWid) testTxs mempty
+            testOpeningCleaning f
+                (\db' -> readTxHistory db' testWid defaultTxSortOrder wholeRange)
+                testTxs mempty
```
Minor remark but, when formatting code multiline, try to avoid putting several arguments on the same line. It reads better to have on argument per line:

```hs
            testOpeningCleaning 
                f
                (\db' -> readTxHistory db' testWid defaultTxSortOrder wholeRange)
                testTxs 
                mempty
```

rule of thumb is really: either they all fit on a line, or there's one argument per line.

--> Coding Standard maybe?

Anviking committed:
Fixup: separate Asc and Dsc "put and read tx history"

This is slightly akward. But I opted for "slightly awkward" instead of
"the wrong solution".

Anviking commented on lib/core/src/Cardano/Wallet/Primitive/Types.hs:
```diff
@@ -326,6 +329,36 @@ instance ToText SortOrder where
+    (maybe True (x <=) high)
+
+-- NOTE: We could imagine replacing 'Range (Just 3) Nothing' with something
+-- more magic like: `(Including 3) ... NoBound`
```
I considered this. Maybe superfluous comment…

KtorZ replied:
> with something more magic like

:grimacing: .... We gotta find a use-case for this. Otherwise, it just introduces more code to maintain and more complexity without justification ^.^

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -526,6 +537,21 @@ generator (Model _ wids) = Just $ frequency $ fmap (fmap At) <$> concat
     genPrivKey = elements ["pk1", "pk2", "pk3"]
 
+    genSortOrder :: Gen SortOrder
+    genSortOrder = QC.elements [Ascending, Descending]
```
This can be replaced with `arbitraryBoundedEnum` from `Test.QuickCheck.Arbitrary`:

```hs
arbitraryBoundedEnum :: (Bounded a, Enum a) => Gen a 
```

jonathanknowles commented on lib/core/test/unit/Cardano/WalletSpec.hs:
```diff
@@ -129,12 +130,15 @@ import Test.Hspec
     , (==>)
     )
+import Test.QuickCheck.Instances.Time
+    ()
```
I would recommend replacing this with `Test.Utils.Time.genUniformTime` to get a wider range of `UTCTime` values. The instance supplied by the package above generates times that are clustered around 1864:

See: https://github.com/input-output-hk/cardano-wallet/wiki/Blackboard#investigate-the-arbitrary-instance-for-utctime-in-testquickcheckinstancestime

jonathanknowles commented on lib/core/test/unit/Cardano/Wallet/DB/StateMachine.hs:
```diff
@@ -512,7 +523,7 @@ generator (Model _ wids) = Just $ frequency $ fmap (fmap At) <$> concat
         , (5, ReadWalletMeta <$> genId')
         , (5, PutTxHistory <$> genId' <*> fmap unGenTxHistory arbitrary)
-        , (5, ReadTxHistory <$> genId')
+        , (5, ReadTxHistory <$> genId' <*> genSortOrder <*> genRange)
```
The `genSortOrder` here can be replaced with `arbitraryBoundedEnum`.

jonathanknowles reviewed (commented):
A couple more comments:

1. `genSortOrder` can reuse `arbitraryBoundedEnum`.
2. The `UTCTime` instance from `quickcheck-instances` can be replaced with `Test.Utils.Time.genUniformTime`, as this gives a wider range of `UTCTime` values (rather than those with dates that are clustered around 1864).

**EDIT**: I've added a couple of commits to address the above comments.

jonathanknowles committed:
Use `UniformTime `in module `Cardano.WalletSpec`.

jonathanknowles committed:
Use `arbitraryBoundedEnum` to avoid repeating the definition of `SortOrder`.
